<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Import peaks: ROADMAP — import_peaks_roadmap • PeakyFinders</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Import peaks: ROADMAP — import_peaks_roadmap"><meta property="og:description" content="Import peaks from the 
NIH Roadmap Epigenomics Mapping Consortium."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PeakyFinders</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/peakyfinders.html">Getting Started</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/PeakyFinders/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import peaks: ROADMAP</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/PeakyFinders/blob/HEAD/R/import_peaks_roadmap.R" class="external-link"><code>R/import_peaks_roadmap.R</code></a></small>
    <div class="hidden name"><code>import_peaks_roadmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Import peaks from the <a href="http://www.roadmapepigenomics.org" class="external-link">
NIH Roadmap Epigenomics Mapping Consortium</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">import_peaks_roadmap</span><span class="op">(</span>
  <span class="va">ids</span>,
  build <span class="op">=</span> <span class="st">"hg19"</span>,
  query_granges <span class="op">=</span> <span class="cn">NULL</span>,
  query_granges_build <span class="op">=</span> <span class="cn">NULL</span>,
  split_chromosomes <span class="op">=</span> <span class="cn">FALSE</span>,
  condense_queries <span class="op">=</span> <span class="cn">TRUE</span>,
  force_new <span class="op">=</span> <span class="cn">FALSE</span>,
  call_peaks_method <span class="op">=</span> <span class="st">"MACSr"</span>,
  cutoff <span class="op">=</span> <span class="cn">NULL</span>,
  searches <span class="op">=</span> <span class="fu"><a href="construct_searches.html">construct_searches</a></span><span class="op">(</span><span class="op">)</span>,
  peaks_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,
  nThread <span class="op">=</span> <span class="fl">1</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ids</dt>
<dd><p>IDs from one of the supported databases. 
IDs can be at any level: file, sample, or experiment.</p></dd>
<dt>query_granges</dt>
<dd><p>[Optional]
<a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object indicating which genomic
regions to extract from each sample.</p></dd>
<dt>query_granges_build</dt>
<dd><p>[Optional]
Genome build that <code>query_granges</code> is aligned to.</p></dd>
<dt>split_chromosomes</dt>
<dd><p>Split single-threaded query 
into multi-threaded query across chromosomes. 
This is can be helpful especially when calling peaks from 
large bigWig/bedGraph files.
The number of threads used is set by the <code>nThread</code> argument.</p></dd>
<dt>condense_queries</dt>
<dd><p>Condense <code>query_granges</code>
by taking the min/max position per chromosome (default: <code>TRUE)</code>.
This helps to reduce the total number of queries, 
which can cause memory allocation problems 
due to repeated calls to the underlying C libraries.</p></dd>
<dt>force_new</dt>
<dd><p>By default, saved results of the same <code>save_path</code> name 
will be imported instead of running queries. However you can override this
by setting <code>force_new</code> to perform new queries regardless and overwrite
 the old <code>save_path</code> file.</p></dd>
<dt>call_peaks_method</dt>
<dd><p>Method to call peaks with:</p><ul><li><p>"MACSr" : Uses <a href="https://github.com/macs3-project/MACS" class="external-link">MACS3</a> 
via <a href="https://rdrr.io/pkg/MACSr/man/bdgpeakcall.html" class="external-link">bdgpeakcall</a>.</p></li>
</ul></dd>
<dt>cutoff</dt>
<dd><p>Cutoff depends on which method you used for score track.
If the file contains pvalue scores from MACS3, score 5 means pvalue 1e-5.
If <code>NULL</code>, a reasonable <code>cutoff</code> value will be inferred 
through a <code>cutoff_analysis</code>.</p></dd>
<dt>searches</dt>
<dd><p>Named list of regex queries.</p></dd>
<dt>peaks_dir</dt>
<dd><p>Directory to save peaks to 
(only used when calling peaks from bedGraph files).</p></dd>
<dt>nThread</dt>
<dd><p>When <code>nThread&gt;1</code>, accelerates file importing 
and peak calling using multi-core parallelisation.</p></dd>
<dt>verbose</dt>
<dd><p>Print messages.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

